alacritty_terminal/src/term/cell.rs

```diff
@@ -29,31 +27,84 @@ bitflags! {
     }
 }
 
-const fn default_extra() -> [char; MAX_ZEROWIDTH_CHARS] {
-    [' '; MAX_ZEROWIDTH_CHARS]
+/// Trait for determining if a reset should be performed.
+pub trait ResetDiscriminant<T> {
+    /// Value based on which equality for the reset will be determined.
+    fn discriminant(&self) -> T;
+}
+
+impl<T: Copy> ResetDiscriminant<T> for T {
+    fn discriminant(&self) -> T {
+        *self
+    }
+}
+
+impl ResetDiscriminant<Color> for Cell {
+    fn discriminant(&self) -> Color {
+        self.bg
+    }
+}
+
+/// Dynamically allocated cell content.
+///
+/// This storage is reserved for cell attributes which are rarely set. This allows reducing the
+/// allocation required ahead of time for every cell, with some additional overhead when the extra
+/// storage is actually required.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+struct CellExtra {
+    zerowidth: Vec<char>,
 }
 
-#[derive(Copy, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+/// Content and attributes of a single cell in the terminal grid.
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 pub struct Cell {
     pub c: char,
     pub fg: Color,
     pub bg: Color,
     pub flags: Flags,
-    #[serde(default = "default_extra")]
-    pub extra: [char; MAX_ZEROWIDTH_CHARS],
+    #[serde(default)]
+    extra: Option<Box<CellExtra>>,
 }
 
 impl Default for Cell {
+    #[inline]
     fn default() -> Cell {
-        Cell::new(' ', Color::Named(NamedColor::Foreground), Color::Named(NamedColor::Background))
+        Cell {
+            c: ' ',
+            bg: Color::Named(NamedColor::Background),
+            fg: Color::Named(NamedColor::Foreground),
+            flags: Flags::empty(),
+            extra: None,
+        }
+    }
+}
+
+impl Cell {
+    /// Zerowidth characters stored in this cell.
+    #[inline]
+    pub fn zerowidth(&self) -> Option<&[char]> {
+        self.extra.as_ref().map(|extra| extra.zerowidth.as_slice())
+    }
+
+    /// Write a new zerowidth character to this cell.
+    #[inline]
+    pub fn push_zerowidth(&mut self, c: char) {
+        self.extra.get_or_insert_with(Default::default).zerowidth.push(c);
+    }
+
+    /// Free all dynamically allocated cell storage.
+    #[inline]
+    pub fn drop_extra(&mut self) {
+        if self.extra.is_some() {
```

Comment:
Why not  always do `self.extra = None`, should be faster than useless branch here?

Reply:
Because it's slower:

![tmp](https://user-images.githubusercontent.com/8886672/98045130-e59ca580-1e1f-11eb-8cca-5d0ee0c39150.png)

![tmp](https://user-images.githubusercontent.com/8886672/98045181-f8af7580-1e1f-11eb-988f-e7612af21421.png)

Reply:
That it's so strange, but whatever. Like `None = None` should be the same...

Reply:
I'm not entirely sure how it is implemented, but it seems to do more than just check for emptiness itself.